tests/test_rpc.c

```diff
@@ -28,100 +28,325 @@
+    nc_rpc_free(rpc);
+
+    /* TODO: NC_RPC_ESTABLISHSUB and the rest */
+}
+
 static void
 test_lock(void **state)
 {
     struct np_test *st = *state;
-    struct nc_session *nc_sess2;
     struct nc_rpc *rpc;
     NC_MSG_TYPE msgtype;
     uint64_t msgid;
     struct lyd_node *envp, *op;
     char *str, *str2;
 
-    /* create another NETCONF session */
-    nc_sess2 = nc_connect_unix(NP_SOCKET_PATH, NULL);
-    assert_non_null(nc_sess2);
-
-    /* lock RPC */
+    /* lock from first session */
     rpc = nc_rpc_lock(NC_DATASTORE_RUNNING);
     assert_non_null(rpc);
 
-    /* send request on session #1 */
-    msgtype = nc_send_rpc(st->nc_sess, rpc, 1000, &msgid);
-    assert_int_equal(msgtype, NC_MSG_RPC);
+    /* send request */
+    assert_int_equal(NC_MSG_RPC, nc_send_rpc(st->nc_sess, rpc, 1000, &msgid));
 
     /* receive reply */
     msgtype = nc_recv_reply(st->nc_sess, rpc, msgid, 2000, &envp, &op);
     assert_int_equal(msgtype, NC_MSG_REPLY);
     assert_null(op);
     assert_string_equal(LYD_NAME(lyd_child(envp)), "ok");
+
     lyd_free_tree(envp);
 
-    /* send request on session #2 */
-    msgtype = nc_send_rpc(nc_sess2, rpc, 1000, &msgid);
+    /* request to lock from another session should fail when lock already */
+    msgtype = nc_send_rpc(st->nc_sess2, rpc, 1000, &msgid);
     assert_int_equal(msgtype, NC_MSG_RPC);
 
-    /* receive reply */
-    msgtype = nc_recv_reply(nc_sess2, rpc, msgid, 2000, &envp, &op);
+    /* recieve reply, should yield error */
+    msgtype = nc_recv_reply(st->nc_sess2, rpc, msgid, 2000, &envp, &op);
     assert_int_equal(msgtype, NC_MSG_REPLY);
     assert_null(op);
+    assert_string_equal(LYD_NAME(lyd_child(envp)), "rpc-error");
     assert_int_equal(LY_SUCCESS, lyd_print_mem(&str, envp, LYD_XML, LYD_PRINT_SHRINK));
-    lyd_free_tree(envp);
 
+    assert_int_not_equal(-1, asprintf(&str2, LOCK_FAIL_TEMPLATE, msgid, nc_session_get_id(st->nc_sess)));
     /* error expected */
-    asprintf(&str2,
-    "<rpc-reply xmlns=\"urn:ietf:params:xml:ns:netconf:base:1.0\" message-id=\"%d\">"
-        "<rpc-error>"
-            "<error-type>protocol</error-type>"
-            "<error-tag>lock-denied</error-tag>"
-            "<error-severity>error</error-severity>"
-            "<error-message lang=\"en\">Access to the requested lock is denied because the lock is currently held by another entity.</error-message>"
-            "<error-info>"
-                "<session-id>1</session-id>"
-            "</error-info>"
-        "</rpc-error>"
-    "</rpc-reply>", (int)msgid);
     assert_string_equal(str, str2);
     free(str);
     free(str2);
 
     nc_rpc_free(rpc);
+    lyd_free_tree(envp);
 
     /* unlock RPC */
     rpc = nc_rpc_unlock(NC_DATASTORE_RUNNING);
-    assert_non_null(rpc);
+    assert_int_equal(NC_MSG_RPC, nc_send_rpc(st->nc_sess, rpc, 1000, &msgid));
+    msgtype = nc_recv_reply(st->nc_sess, rpc, msgid, 2000, &envp, &op);
+    assert_int_equal(msgtype, NC_MSG_REPLY);
```

Comment:
You should probably stick to either the assert, expected return, and function call on one line or put the assert on a separate line, not mix it. I prefer having the assert on separate line because otherwise there is a bit too much clutter.